File: MC_Graph simulation/mc_AveNearNeighborDeg.m
function [aveK,k_bin]=mc_AveNearNeighborDeg(Adj,bin,inout);
%gives a list of k-mean degree connectivity for the network for successive
%k=0,1,2....
% The average degree connectivity is also known as average nearest neighbor degree.
% The k-average degree connectivity is the average of the mean neighbor
% degrees of vertices of degree k
% 
%       Input: Adj - adjacencey matrix (also weighted and directed)
%                bin = subdivision of the degree interval on which find the
%                          average connectivity
%                inout = choose if "in", "out" or "total"
% it works also for weighted and directed networks
%@F.Vanni2023 

if nargin<3
    inout='total'
end


if strcmpi(inout,'in');Adj=Adj'; ave='in';
elseif strcmpi(inout,'out');Adj=Adj; ave='in';
else ave='total';
end

[deg,indeg,outdeg]=degrees(Adj);

switch ave
    case 'in'
        degr=max(indeg);
        k_bin=linspace(0,degr,bin);
        aveK=zeros(1,length(k_bin));
        k=indeg;
        knn=aveNeighborDegWW(Adj,'in');

        for i=1:length(k_bin)-1
           ix=(k>k_bin(i) & k<=k_bin(i+1) );
           aveK(i)=mean(knn(ix));
        end
        
        
    case 'out'
        degr=length(outdeg);
        k_bin=linspace(0,degr,bin);
        aveK=zeros(1,length(k_bin));
        k=outdeg;
        knn=aveNeighborDegWW(Adj,'out');

        for i=1:length(k_bin)-1
           ix=(k>k_bin(i) & k<=k_bin(i+1) );
           aveK(i)=mean(knn(ix));
        end
        
    case 'total'
        degr=max(deg);
        k_bin=linspace(0,degr,bin);
        aveK=zeros(1,length(k_bin));
        k=deg;
        knn=aveNeighborDegWW(Adj,'total');

        for i=1:length(k_bin)-1
           ix=(k>k_bin(i) & k<=k_bin(i+1) );
           aveK(i)=mean(knn(ix));
        end

end




function ave_n_deg=aveNeighborDegWW(adj,inout)
%  Average weighted and directed neighbor degree


ave_n_deg=zeros(1,length(adj));   % initialize output vector

A=double(adj>0); % adjacency matrix
W=adj; % weight matrix



if nargin<2
    inout='total';
end

switch lower(inout)
    case 'in'
        adj=adj;
        deg = sum(adj);
    case 'out'
        adj=transpose(adj);
       deg = sum(adj);
    case 'total'
        [deg,~,~]=degrees(adj);
end


for i=1:length(adj)  % across all nodes
  
  neigh=kneighbors(adj,i,1);  % neighbors of i, one link away
  peso=W(i,neigh);
  if isempty(neigh); ave_n_deg(i)=0; continue; end
  stren(i)=sum(peso);
  ave_n_deg(i)=sum(deg(neigh).*peso)/(stren(i));
end

File: MC_Graph simulation/mc_AveNearNeighborDeg.py
import numpy as np

def mc_AveNearNeighborDeg(Adj, bin, inout='total'):
    # Gives a list of k-mean degree connectivity for the network for successive k=0,1,2, ...
    # The average degree connectivity is also known as the average nearest neighbor degree.
    # The k-average degree connectivity is the average of the mean neighbor degrees of vertices of degree k.
    # Input:
    # Adj - adjacency matrix (also weighted and directed)
    # bin - subdivision of the degree interval on which to find the average connectivity
    # inout - choose if "in", "out" or "total"

    if inout.lower() == 'in':
        Adj = Adj.T  # Transpose the adjacency matrix for "in" case
        ave = 'in'
    elif inout.lower() == 'out':
        ave = 'in'
    else:
        ave = 'total'

    deg = np.sum(Adj, axis=1)
    k_bin = np.linspace(0, np.max(deg), bin)
    aveK = np.zeros(bin - 1)
    k = deg
    knn = aveNeighborDegWW(Adj, 'in')

    for i in range(bin - 1):
        ix = np.logical_and(k > k_bin[i], k <= k_bin[i + 1])
        aveK[i] = np.mean(knn[ix])

    return aveK, k_bin


def aveNeighborDegWW(adj, inout='total'):
    # Average weighted and directed neighbor degree

    ave_n_deg = np.zeros(len(adj))  # initialize output vector

    adj = np.double(adj > 0)  # adjacency matrix
    W = adj  # weight matrix

    deg = np.sum(adj, axis=1)  # calculate the degree of each node

    for i in range(len(adj)):  # across all nodes
        neigh = np.where(adj[i] == 1)[0]  # neighbors of i, one link away
        peso = W[i, neigh]
        if len(neigh) == 0:
            ave_n_deg[i] = 0
            continue
        stren = np.sum(peso)
        ave_n_deg[i] = np.sum(deg[neigh] * peso) / stren

    return ave_n_deg

# Rest of the code...
# ...

# Example usage
# Replace Adj with your adjacency matrix (numpy array)
# bin: the number of bins for calculating the average connectivity
# inout: 'in', 'out', or 'total' for the desired type of neighbor degree
aveK, k_bin = mc_AveNearNeighborDeg(Adj, bin, inout)

# Rest of the code...
# ...

